README.md
# **pkf_trackpoint-qt** v2.0.1
Qt application for configuring a TrackPoint

## Description
This program will allows you to change the speed, sensitivity, and press_to_select settings of the TrackPoint on ThinkPads running GNU/Linux ***(See "Limitations" below)***   

## Credits
This program would not be possible if it was not for the amazing wealth of knowledge from www.thinkwiki.org.  Please consider contributing or donating to this page.
Acknowledgment is also due to www.gnu.org and www.fsf.org for all their amazing work with the GNU license and Free Software.  Please consider donating or becoming a member of the FSF!

## Usage
### Modes
- Normal: Run program as root or by running it with "sudo" for use of all features.  Alternatively, users can set ownership to "root" and enable the "setuid" bit on file permissions, as follows: *NOTE: GNOME does not generally support using the "setuid" executing with root privileges*
```
chown root pkf_trackpoint-qt
chmod u+s pkf_trackpoint-qt
```
- ReadOnly: Run program as normal user to view current TrackPoint settings
- Test:  Run program pointing to a testing directory.  Allows normal users to test all functionality and allows root to use without changing any system settings     
*NOTE: All files generated by "Test" mode are deleted when the user exists the program*

### Operation
- To change one or more settings of the TrackPoint, enter the appropriate value in the text field and click "Apply".  Text fields left blank are ignored.
- To enable current settings to persist reboots, check the "Enable Persistence" checkbox
- To disable persistence (and remove any files created for persistence), un-check the "Enable Persistence" checkbox
- To reset the TrackPoint settings to default values, click "Defaults".
- To enable scrolling on a distribution that uses "Evdev" for device management, click the checkbox at the button of the screen titled "Enable Scrolling"  
*NOTE: This checkbox will only be visible if your distribution is using "Evdev".  To determine what your distribution is using, run the following*
```bash
xinput list-props "TPPS/2 IBM TrackPoint" | grep Evdev
```   

## Limitations
All Limitations will be addressed in future versions
- Persistence currently only works for systemd

### Disclaimer...
This program was written by a novice who is just learning C++ and Qt.  As such, the author is aware that there is much room for improvement in the code and appreciates any suggestions for cleaning it up or making it more efficient.  The author can be reached at "afogl001@members.fsf.org".    

### Planned Improvements
- Allow Evdev users to choose wither to initiate scrolling
- Fix minor offset of labels
- Do now allow user to un-check persist if not root
- Add initd support for persistence
- Allow user to run and prompt for escalated privileges
- Add menu items for "Close", "Help", and "About".
- Clean up code by
  - Making better use of functions and parameters (e.x., make "install" functions uninstall as well)
  - Leverage variables to reduce repetitious text/values
  - Find better way of creating persistence files (instead of just streaming line-by-line)
  - Consider breaking out some code from main and put in other classes
- Create binaries
- Create .deb and .rpm packages

### **Code overview**
#### Functions
+ setSettingsPath() = Sets path for TrackPoint settings (aka, speed, sensitivity, and press_to_select)
  - Runs directory iterator loop to determine correct TrackPoint settings path (starts with default value of  *testPath*)
  - If no paths match those tested, show alert window and run **setTestMode** with "1", putting the user in "Test" mode.
+ applySettings(QString settingsFile, QString settingsValue) = Applies any specified TrackPoint setting
  - If *settingsValue* not NULL, applies that value to the *settingsFile*
+ validateSettings() = Configure validators for text edit boxes
  - Create QValidators for values "1-255" and "0-1"
  - Set speed and sensitivity edit texts boxes with "1-255" validator
  - set press_to_select edit text box to "0-1" validator
+ displaySettings() = Displays current TrackPoint setting values
  - Checks existence of a "speed" file, throwing warning error to user if it doesn't exist or passing value to appropriate label if it does
  - Checks existence of a "sensitivity" file, throwing warning error to user if it doesn't exist or passing value to appropriate label if it does
  - Checks existence of a "press_to_select" file, throwing warning error to user if it doesn't exist or passing value to appropriate label if it does
+ setPersistPaths() = Sets path for initialization settings, used for configuring persistent TrackPoint settings
  - Runs conditional to determine what path to use based on existence (starts with default value of *testPath*) and sets values as follows...
    - If "speed" file exist in "/tmp/pkf_trackpoint", set *trackpointSHPath* and *initPath* to "/tmp/pkf_trackpoint"
    - If "/etc/systemd/system" exist (systemd), set *trackpointSHPath* to "/usr/bin" and *initPath* to "/etc/systemd/system"
    - If no paths match, set *trackpointSHPath* and *initPath* to "N/A"
  - Run conditional to check existence of persistent script ("trackpoint.sh"), checking the persist checkbox if existing or unchecking otherwise
+ checkMode() = Set's *initCommand* and *label_info* based on current mode
  - Gets effective user of running application
  - If root and not in test mode, *initCommand* = "systemctl" and *label_info* = cleared ("")
  - Else, if in test mode, *initCommand* = "echo systemctl" and *label_info* = Test mode info
  - Else, *initCommand* = "echo systemctl" and *label_info* = ReadOnly mode info
  - Send *effectiveUser* and *initCommand* to qDebug
+ setTestMode(int testMode) = Takes [0,1,3] and sets mode accordingly
  - Sets *testStatus* = *testMode*
  - Sends passed *testMode* option to qDebug
  - Sets *testPath*
  - If *testMode* = "1" (On)
    - Create directory specified by *testPath*
    - Create "speed", "sensitivity", and "press_to_select" files in *testPath* with contents "Test"
    - Run **setSettingsPath()**, **setPersistPaths()**, **displaySettings()**, and **checkMode()**
  - If *testMode* = "0" (Off)
    - Recursively remove test directory (*testPath*)
    - Run **setSettingsPath()**, **setPersistPaths()**, **displaySettings()**, and **checkMode()**
  - If *testMode* = "3" (Cleanup).  This is used to remove all testing files but without run other functions that will cause a segmentation fault if run at application exit
    - Recursively remove test directory (*testPath*).  
+ on_button_Apply_clicked() = Applies any specified settings
  - Runs **applySettings(<settingsFile>, <ui_settings_edit>)** for all three settings (blank or NULL values are handled within called function)
  - Runs **displaySettings()**
  - If the persistence checkbox is checked, runs **installTrackpointSH()**
  - Clears all values in the text edit boxes
+ on_button_Defaults_clicked() = Sets TrackPoint settings to default values
  - Sets "speed" to "97"
  - Sets "sensitivity" to "128"
  - Sets "press_to_select" to "0"
  - Runs **displaySettings()**
  - If persist checkbox is checked, runs **installTrackpointSH()**
+ on_check_Persist_clicked() = Toggles settings persistence
  - If checkbox is checked when it's clicked...
     - If *trackpointSHPath* is "N/A", critical alert user no initialization and uncheck checkbox
     - Else, run **installTrackpointSH()**, **installTrackpointService()**, **installTrackpointTimer()**, and run initialization commands to configure installed persistent settings
   - If checkbox is not checked when it's clicked...
     - If *trackpointSHPath* is "N/A", do nothing
     - Else...
       - Run initialization commands to to stop/disable persistent settings
       - Delete generated files for persistence
       - Run command to reload initialization system
       - Run **applySettings(<settingsFile>, <ui_settings_label>)** for all three settings to maintain current settings (defaults can be set by clicking "Defaults" button)
       - Run **displaySettings()**
+ on_actionToggle_Test_Mode_triggered = Toggles test mode
  - If "speed" file exist in *testPath*, run **setTestMode(0)**  (disable test mode)
  - Else, run **setTestMode(1)** (enable test mode)
+ installTrackpointSH() = Generates "trackpoint.sh" script for persistence configuration
+ installTrackpointService() = Generates "trackpoint.service" for persistence configuration
+ installTrackpointTimer() = Generates "trackpoint.timer" for persistence configuration

#### Variables (Local)
+ effectiveUser:auto = In **checkMode()**, stores uid of user who is running the application
+ testPath:= In **setTestMode()**, path used for storing all outputed files created by application when in Test mode
+ testMode: = In **setTestMode()**, stores option passed to function

#### Variables (Global)
+ settingPath:QString = Path to TrackPoint settings (e.x., "/sys/devices/platform/i8042/serio1/serio2")
+ trackpointSHPath:QString = Path for trackpoint.sh script (used for persistence configuration)
+ initPath:QString = Path for initialization system configurations (e.x., "/usr/bin").  Used for persistence configuration
+ initCommand:QString = Command for interaction with initialization system (e.x., "systemcdtl").  Used for persistence configuration
+ testPath:QString = Path for all generated files while in "Test" mode (e.x., "/tmp/pkf_trackpoint")
+ testStatus:int = Status for "Test Mode"
  - "0" = off
  - "1" = on
  - "3" = cleanup during application shutdown (similar to "0" but does not run other functions)

#### Process Flow
The main loop will...
1. Set the path for TrackPoint settings by running "setSettingsPath", which will...
  - If in "Test" mode **(see "Modes" above)**, set path to the testing directory "/tmp/pkf_trackpoint"
  - Otherwise, as long as a TrackPoint settings directory is detected (depends on existence of trackpad), set path appropriately
  - If no TrackPoint setting directory is detected, alerts user and exits program
1. Set the path for persistence files by running "setPersistPaths", which will...
  - If in "Test" mode **(see "Modes" above)**, set path to the testing directory "/tmp/pkf_trackpoint"
  - Otherwise, as long as an initialization system is detected (currently only supports systemd), set path appropriately
  - If no initialization setting directory is detected, sets path to "N/A", which will prevent the user from checking the "Enable Persistence" checkbox.
1. Check whether the user is root or not by running "checkUser", which will...
  - If effective user is root, sets initialization command to actual command (i.e., "systemctl" for systemd)
  -  If effective user is anyone other than root, sets initialization command with prefixed "echo" (i.e., "echo systemctl" for systemd) and displays text in Info label for the user to run as root or with sudo to make changes.
1. Set validators for the text input fields by running "validateSettings" which will...
  - Set the input text boxes for speed and sensitivity only accept values 1-255 (though user still permitted to enter "0")
  - Set the input text box for press_to_select to only accept values 0-1
1. Display the current TrackPoint settings by running "displaySettings", which will...
  - Read TrackPoint configuration settings and print them to the appropriate labels.
1. Check the "Enable Persistence" checkbox if it detects persistence has been enabled.

User can then...
1. Enter values into text boxes and click "Apply", which will...
  - Run "applySettings" for all three settings, passing in the appropriate settings path and text of respective text box, which will in turn apply the setting only if the text box was not empty.
  - Update the displayed values by running "displaySettings"
  - Run "installTrackpointSH" **(covered below in "Enable Persistence...")**, only if persistence is enabled, to apply changes made
  - Clear all text boxes.
1. Reset TrackPoint settings to defaults by clicking "Defaults", which will...
  - Run "applySettings" for all three settings, passing in the appropriate settings path and default value, which will in turn apply all settings to default values
  - Update the displayed values by running "displaySettings"
  - Run "installTrackpointSH" **(covered below in "Enable Persistence...")** only if persistence is enabled, to apply changes made   
1. Enable Persistence by checking the "Enable Persistence" which will...
  - If status is changing to "checked"...
    - If no initialization system is detected, alert user and un-checks the checkbox
    - Else, runs functions to create files for persistence, then executes commands to run and enable persistence via the initialization system
  - If status is changing to "unchecked"...
    - If no initialization system is detected, notification passed to console that nothing is changing.
    - Else, commands run to disable and stop persistence, then all files created for persistence are removed, then initialization is reloaded if required.
1. Toggle "Test" mode by clicking on "Advanced -> Toggle "Test" mode", which will...
  - If currently not in "Test" mode...
    - Create directory and files for test mode
    - Update information label that user is now in "Test" mode
    - Reruns "setSettingsPath", "setPersistPaths", "displaySettings", and "checkMode"
    - All interaction with application will be applied to files within the created "Test" directory
  - If currently in "Test" mode...
    - Recursively removes the "Test" directory and all it's contents
    - Updates information label to "ReadOnly" for regular users or "" for root user
    - Reruns "setSettingsPath", "setPersistPaths", "displaySettings", and "checkMode"
    - User is now in "ReadOnly" mode if regular user, or "Normal" if root user


CHANGE_LOG.md
# **v2.0**
## v2.0.1
  - Other
    + Corrections to README and acknowledgments to very important resources.
## v2.0.0
  - Features
    + Add "Enable scrolling" for distros using "Evdev" device manager
    + Add window icon
  - Other
    + Add variable "runMode" to track if running in standard mode (with root privileges), "Test" mode, or "ReadOnly" mode
    + Updated some misc. comments in code
    + Add "CONFIG += console" to project file in order to view qDebug info on CentOS and OpenSUSE builds of Qt Creator.
    + Removed statusBar and mainToolBar as they were not being used and took up screen space.
  - Bugs
    + Introduced unsetting scrolling settings bug.  If 1) running in standard mode, 2) toggling "Test" mode off and on again, 3) distro running Evdev, and 4) user has enabled scrolling, then scrolling will be unset.  Remediation is to reset scrolling (by click the appropriate checkbox) before exiting.  While there are several methods of addressing this bug, it's numerous prerequisites and ease of remediation don't seem to warrant introducing additional code to address.  This may be re-investigated later or if a sufficiently simple method of address becomes apparent.
    + Introduced delay in starting pkf_trackpoint-qt.  If 1) running a distro that uses Evdev, 2) uses a slow and large HDD, and 3) does not have scrolling set, then pkf_trackpoint-qt may take up to 35 seconds to start (running CentOS on X220 w/ 5400rpm HDD).  This is due to the "checkDeviceMgr" function recursively searching possible locations of the "xorg.conf.d" directory with the "90-trackpoint.conf" file.  Added addressing this under the "Planned Improvements" section of README but the likelihood and impact of bug does not warrant withholding release of existing version.         


# v1.2
## v1.2.0
  - Features
    + Add dynamic discovery of TrackPoint settings path
    + Add condition if TrackPoint not discovered, open in "Test" mode
  - Other
    + Clean up some conditionals with explicit brackets
# v1.1
## v1.1.0
  - Features
    + Add support for KDE Neon
    + Add menu entry for toggling "Test" mode
    + Enable root user to use "Test" mode
  - Fixes
    + Fix typo where un-checking "Set Persistence" did not remove systemd files
    + Fix plethora of typos in trackpoint.sh file
    + Fix typo/misspelling of several instances of "sensitivity"
    + Current settings maintained when un-checking "Set Persistence"
  - Other
   + Remove dynamic path assignment in trackpoint.sh.  Provided fail-safe if moved to other host but was largely unnecessary and redundant
   + Dynamically add value for trackpoint.service "ConditionPathExists" instead of statically listing all possibilities
   + Add status check for "Persist" checkbox in "setPersistPaths"
   + Update README
   + Remove extraneous/unused variables for specific setting values (e.x., "speedPath")

# v1.0
Initial functional release


# v0.0.1
Initial commit of non-functional draft


mainwindow.h
#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include <QMainWindow>
#include <QFile>
#include <QDir>
#include <QDirIterator>
#include <QIODevice>
#include <QTextStream>
#include <QMessageBox>
#include <QDebug>
#include <QMessageBox>
#include <QProcess>

#include <unistd.h>  // For "getuid" and "geteuid"
#include <QIntValidator>  // Added since QIntValidator was moved to it's own header

namespace Ui {
class MainWindow;
}

class MainWindow : public QMainWindow
{
    Q_OBJECT

public:
    explicit MainWindow(QWidget *parent = 0);
    ~MainWindow();
    void setSettingsPath();
    void applySettings(QString settingFile, QString settingValue);
    void validateSettings();
    void displaySettings();
    void origionalFunction();
    void installTrackpointSH();
    void installTrackpointService();
    void installTrackpointTimer();
    void setPersistPaths();
    void checkMode();
    void setTestMode(int testMode);
    void checkDeviceMgr();

private slots:
    void on_button_Apply_clicked();

    void on_button_Defaults_clicked();

    void on_check_Persist_clicked();

    void on_actionToggle_Test_Mode_triggered();

    void on_check_Scrolling_stateChanged(int arg1);

private:
    Ui::MainWindow *ui;
    QString settingPath;
    QString trackpointSHPath;
    QString initPath;
    QString initCommand;
    QString testPath;
    QString testXorgPath;
    int testStatus;
    QString xorgLocation;
    QString xinputCommand;
    int runMode;
};

#endif // MAINWINDOW_H


mainwindow.cpp
#include "mainwindow.h"
#include "ui_mainwindow.h"

MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    checkDeviceMgr();
    setSettingsPath();
    setPersistPaths();
    checkMode();
    validateSettings();
    displaySettings();
}

MainWindow::~MainWindow()
{
    delete ui;
    setTestMode(3);
}

void MainWindow::checkDeviceMgr()
{
    QProcess eventDevice;
    QString exec = "xinput";
    QStringList params;
     params << "list-props" << "TPPS/2 IBM TrackPoint";

    eventDevice.start(exec, params);
    eventDevice.waitForFinished(); // sets current thread to sleep and waits for process to end
    QString output(eventDevice.readAllStandardOutput());
    qDebug() << "Using Evdev: " << output.contains("Evdev");
    if (output.contains("Evdev")) {
       ui->check_Scrolling->setVisible(true);
       // Determine and set correct path for X11 xorg.conf.d
       QStringList xOrgDirList;  // Declare list for directories to search
       xOrgDirList << "/tmp" <<  "/usr" << "/etc";  // Specify directories to search
       bool fileFound = false;

       foreach (QString element, xOrgDirList) {
        QDirIterator xorg_check(element, QStringList() << "xorg.conf.d", QDir::NoFilter, QDirIterator::Subdirectories);
        while (xorg_check.hasNext() && fileFound == false) {  // Search recursivly for "xorg.conf.d"
            qDebug() << "Checking for xorg.conf.d, found: " << xorg_check.next();  // For some reason, this is needed or eles the infinite loop
            xorgLocation = xorg_check.fileInfo().absoluteFilePath();  // Assign "xorg.conf.d" directory to variable
            qDebug() << "Using xorg.conf.d directory: " << xorgLocation;

            QFileInfo fileInfoXorg(xorgLocation+"/90-trackpoint.conf");  // Set object for "90-trackpoint.conf" file
            qDebug() << "Scrolling file: " << fileInfoXorg.exists();
            if (fileInfoXorg.exists()) {  // If "90-trackpoint.conf"...
                ui->check_Scrolling->setChecked(1);
                fileFound = true;
           }
            else ui->check_Scrolling->setChecked(0);
        }
       }
    }
    else ui->check_Scrolling->setVisible(false);
}

void MainWindow::checkMode()
{
        auto effectiveUser = geteuid();

        if (effectiveUser == 0 && testStatus!=1) {
            initCommand = "systemctl";
            xinputCommand = "xinput";
            ui->label_Info->setText("");
            runMode=0;
        }
        else if (testStatus==1) {
            initCommand = "echo systemctl";
            xinputCommand = "echo xinput";
            ui->label_Info->setText("Running in \"Test\" Mode");
            runMode=1;
        }
        else {
            initCommand = "echo systemctl";
            xinputCommand = "echo xinput";
            ui->label_Info->setText("Running in \"ReadOnly\" Mode:  Run with \"sudo\" or as root to change to TrackPoint settings");
            runMode=2;
        }
        qDebug() << "Effective User = " << effectiveUser;
        qDebug() << "initCommand = " << initCommand;
        qDebug() << "xinputCommand = " << xinputCommand;
}

void MainWindow::setTestMode(int testMode)
{
    testStatus=testMode;
    qDebug() << "Setting \"Test Mode\" to" << testMode;
    testPath = "/tmp/pkf_trackpoint";
    QDir dir(testPath);
    if (testMode==1) {
        qDebug() << "Test path = " << testPath;
        dir.mkdir(testPath);
        QProcess::execute("mkdir " + testPath + "/xorg.conf.d");
        QProcess::execute("touch " + testPath + "/xorg.conf.d/90-trackpoint.conf");

        QFile testSpeed(testPath + "/speed");
        testSpeed.open(QIODevice::WriteOnly | QIODevice::Truncate);
        QTextStream streamSpeed(&testSpeed);
        streamSpeed << "Test";
        testSpeed.close();

        QFile testSensitivity(testPath + "/sensitivity");
        testSensitivity.open(QIODevice::WriteOnly | QIODevice::Truncate);
        QTextStream streamSensitivity(&testSensitivity);
        streamSensitivity << "Test";
        testSensitivity.close();

        QFile testPress_to_select(testPath + "/press_to_select");
        testPress_to_select.open(QIODevice::WriteOnly | QIODevice::Truncate);
        QTextStream streamPress_to_select(&testPress_to_select);
        streamPress_to_select << "Test";
        testPress_to_select.close();

        setSettingsPath();
        setPersistPaths();
        displaySettings();
        checkMode();
        checkDeviceMgr();
    }
    else if (testMode==0) {
        qDebug() << "Removong " << testPath;
        dir.removeRecursively();

        setSettingsPath();
        setPersistPaths();
        displaySettings();
        checkMode();
        checkDeviceMgr();
    }
    else if (testMode==3) {
        qDebug() << "Removong " << testPath << " if it exists";
        dir.removeRecursively();
    }
    else
        qDebug() << "Use options 1-3 for setting Test Mode";
}

void MainWindow::on_button_Apply_clicked()
{
    applySettings(settingPath + "/speed", ui->edit_Speed->text());
    applySettings(settingPath + "/sensitivity", ui->edit_Sensitivity->text());
    applySettings(settingPath + "/press_to_select", ui->edit_Press_To_Select->text());

    displaySettings();

    if (ui->check_Persist->isChecked())
        installTrackpointSH();

    ui->edit_Speed->clear();
    ui->edit_Sensitivity->clear();
    ui->edit_Press_To_Select->clear();
}

void MainWindow::setSettingsPath()
{
  QStringList dirList;  // Declare list for directories to search
  dirList << "/sys/devices/" << "/tmp/";  // Specify directories to search

  foreach (QString element, dirList) {
   QDirIterator file_check(element, QStringList() << "press_to_select", QDir::NoFilter, QDirIterator::Subdirectories);
   while (file_check.hasNext()) {  // Search recursivly for "press_to_select"
       qDebug() << "Checking for TrackPoint settings, found:  " << file_check.next();  // For some reason, this is needed or eles the infinite loop
       QString fileLocation = file_check.fileInfo().path();  // Assign directory containing "press_to_selectd" to variable
       QFileInfo fileInfoSpeed(fileLocation+"/speed");  // Set object for "speed" file
       QFileInfo fileInfoSensitivity(fileLocation+"/sensitivity");  // Set object for "sensitivity" file
       if (fileInfoSpeed.exists()) {  // If "speed" file exist...
          if (fileInfoSensitivity.exists())  { // and "sensitivity" file exist...
              settingPath = fileLocation;  // set the discovered path to "settingPath"
              qDebug() << "Using TrackPoint settings at: " << settingPath;
          } else {
              QMessageBox::warning(this, "Alert", "No TrackPoint detected!");
              setTestMode(1);
              }
       }

   }
  }
}

void MainWindow::setPersistPaths()
{
    if (QFile::exists("/tmp/pkf_trackpoint/speed")) {  // Make directory "/tmp/pkf_trackpoint" with "speed", "sensitivity", and "press_to_select" if you'd like to test
        trackpointSHPath = "/tmp/pkf_trackpoint";
        initPath = "/tmp/pkf_trackpoint";
    }
    else if (QFile::exists("/etc/systemd/system")) { // If distro using systemd
        trackpointSHPath = "/usr/bin";
        initPath = "/etc/systemd/system";
    }
    else {
        trackpointSHPath = "N/A";
        initPath = "N/A";
    }

    if (QFile::exists(trackpointSHPath + "/trackpoint.sh"))
        ui->check_Persist->setChecked(1);
    else
        ui->check_Persist->setChecked(0);

     qDebug() << "Will place trackpoint.sh in " << trackpointSHPath;
     qDebug() << "Will place initalization files in" << initPath;
     qDebug() << "Persist check-box set to " << ui->check_Persist->checkState();
}

void MainWindow::applySettings(QString settingFile, QString settingValue)
{
    QFile fileSet(settingFile);
    if (settingValue!=NULL) {  // Only update setting if the lineedit is not NULL
        if (fileSet.open(QIODevice::ReadWrite | QIODevice::Truncate)) {   // "QIODevice::Truncate" deletes the content of the file!
            QTextStream stream(&fileSet);
            stream << settingValue;
        }
        fileSet.close();
    }
}

void MainWindow::validateSettings()
{
    QValidator *validateSpeedSensitivity = new QIntValidator(1,255,this);
    QValidator *validatePress_To_Select = new QIntValidator(0,1,this);
    ui->edit_Speed->setValidator(validateSpeedSensitivity);
    ui->edit_Sensitivity->setValidator(validateSpeedSensitivity);
    ui->edit_Press_To_Select->setValidator(validatePress_To_Select);
}

void MainWindow::displaySettings()
{
    QFile speedFile(settingPath + "/speed");
    if(!speedFile.open(QIODevice::ReadOnly)) {
        QMessageBox::warning(0, "Warning", speedFile.errorString());
    }
    ui->label_Speed->setText(speedFile.readAll());
    speedFile.close();

    QFile sensitivityFile(settingPath + "/sensitivity");
    if(!sensitivityFile.open(QIODevice::ReadOnly)) {
        QMessageBox::warning(0, "Warning", sensitivityFile.errorString());
    }
    ui->label_Sensitivity->setText(sensitivityFile.readAll());
    sensitivityFile.close();

    QFile press_to_selectFile(settingPath + "/press_to_select");
    if(!press_to_selectFile.open(QIODevice::ReadOnly)) {
        QMessageBox::warning(0, "Warning", press_to_selectFile.errorString());
    }
    ui->label_PressToSelect->setText(press_to_selectFile.readAll());
    press_to_selectFile.close();
}

void MainWindow::on_button_Defaults_clicked()
{
    applySettings(settingPath + "/speed", "97");
    applySettings(settingPath + "/sensitivity", "128");
    applySettings(settingPath + "/press_to_select", "0");

    displaySettings();

    if (ui->check_Persist->isChecked())
        installTrackpointSH();
}

void MainWindow::on_check_Persist_clicked()
{
    if (ui->check_Persist->isChecked()){
        if (trackpointSHPath=="N/A") {
            QMessageBox::critical(this, "Alert", "Initialization system not detected!");
            ui->check_Persist->setChecked(0);
        }
         else {
            installTrackpointSH();
            installTrackpointService();
            installTrackpointTimer();

            QProcess::execute(initCommand + " daemon-reload");
            QProcess::execute(initCommand + " start trackpoint");
            QProcess::execute(initCommand + " enable trackpoint.timer");
         }
    }

    if (!ui->check_Persist->isChecked()){
        if (trackpointSHPath=="N/A")
            qDebug() << "No initalization system detected so no change made by unchecking";
        else {
            QProcess::execute(initCommand + " disable trackpoint.timer");
            QProcess::execute(initCommand + " stop trackpoint");

            QFile trackpointSH(trackpointSHPath + "/trackpoint.sh");
            trackpointSH.remove();
            QFile trackpointService(initPath + "/trackpoint.service");
            trackpointService.remove();
            QFile trackpointTimer(initPath + "/trackpoint.timer");
            trackpointTimer.remove();

            QProcess::execute(initCommand + " daemon-reload");

            // Reset TrackPoint settings to "current" values (stopping trackpoint.sh sets values to default)
            applySettings(settingPath + "/speed", ui->label_Speed->text());
            applySettings(settingPath + "/sensitivity", ui->label_Sensitivity->text());
            applySettings(settingPath + "/press_to_select", ui->label_PressToSelect->text());
        }
    }
    displaySettings();
}

void MainWindow::on_actionToggle_Test_Mode_triggered()
{
    if (QFile::exists(testPath + "/speed"))
        setTestMode(0);
    else
        setTestMode(1);
}

void MainWindow::installTrackpointSH()
{
    QFile trackpointSH(trackpointSHPath + "/trackpoint.sh");
    trackpointSH.open(QIODevice::WriteOnly | QIODevice::Truncate);
    QTextStream stream(&trackpointSH);
    stream << "#!/bin/bash\n\n";
    stream << "## Determine and set correct path for trackpoint\n";
    stream << "if [ -f " << settingPath << "/speed ];\n";
    stream << "then\n";
    stream << " vTrackpointPath=" << settingPath << "\n";
    stream << "else\n";
    stream << "  echo \"Trackpoint not detected\"\n";
    stream << "  exit 200\n";
    stream << "fi\n\n";
    stream << "## Set trackpoint setting variable values\n";
    stream << "vSensitivity=" << ui->label_Sensitivity->text() << "\n";
    stream << "vSpeed=" << ui->label_Speed->text() << "\n";
    stream << "vPress_to_Select=" << ui->label_PressToSelect->text() << "\n\n";
    stream << "### Apply settings based on option \"start\" or \"stop\" being passed\n";
    stream << "case $1 in\n";
    stream << "  [Ss]tart )\n";
    stream << "    echo -n $vSensitivity > $vTrackpointPath/sensitivity\n";
    stream << "    echo -n $vSpeed > $vTrackpointPath/speed\n";
    stream << "    echo -n $vPress_to_Select > $vTrackpointPath/press_to_select\n";
    stream << "  ;;\n\n";
    stream << "  [Ss]top )\n";
    stream << "    echo 128 > $vTrackpointPath/sensitivity\n";
    stream << "    echo 97 > $vTrackpointPath/speed\n";
    stream << "    echo 0 > $vTrackpointPath/press_to_select\n";
    stream << "  ;;\n\n";
    stream << "  * )\n";
    stream << "    echo \"Usage: {start|stop}\"\n";
    stream << "esac";
    trackpointSH.close();
    trackpointSH.setPermissions(QFileDevice::ReadOwner | QFileDevice::WriteOwner | QFileDevice::ExeOwner | QFileDevice::ReadGroup | QFileDevice::ReadOther);
}

void MainWindow::installTrackpointService()
{
    QFile trackpointService(initPath + "/trackpoint.service");
    trackpointService.open(QIODevice::WriteOnly | QIODevice::Truncate);
    QTextStream stream(&trackpointService);
    stream << "[Unit]\n";
    stream << "Description=Trackpoint Configuration for systemd\n";
    stream << "## Only using \"speed\" file as also using \"sensitivity\" and \"press_to_select\" would be redundant\n";
    stream << "ConditionPathExists=" << settingPath << "/speed\n\n";
    stream << "[Service]\n";
    stream << "Type=oneshot\n";
    stream << "RemainAfterExit=yes\n";
    stream << "ExecStart=" << trackpointSHPath << "/trackpoint.sh start\n";
    stream << "ExecStop=" << trackpointSHPath << "/trackpoint.sh stop\n\n";
    stream << "[Install]\n";
    stream << "WantedBy=graphical.target";
    trackpointService.close();
}

void MainWindow::installTrackpointTimer()
{
    QFile trackpointTimer(initPath + "/trackpoint.timer");
    trackpointTimer.open(QIODevice::WriteOnly | QIODevice::Truncate);
    QTextStream stream(&trackpointTimer);
    stream << "[Unit]\n";
    stream << "Description=Activates trackpoint after boot\n\n";
    stream << "[Timer]\n";
    stream << "OnActiveSec=15\n";
    stream << "Persistent=true\n\n";
    stream << "[Install]\n";
    stream << "WantedBy=timers.target";
    trackpointTimer.close();
}

void MainWindow::on_check_Scrolling_stateChanged(int arg1)
{
    QFile trackpointEvdevConf(xorgLocation + "/90-trackpoint.conf");

    QProcess runXinput;
    QStringList wheelEmuArgs;
    QStringList wheelButtonArgs;
    QStringList wheelAxisArgs;


    if (ui->check_Scrolling->isChecked()) {
        qDebug() << "check_Scrolling: Checked";
        wheelEmuArgs<< "set-prop" << "TPPS/2 IBM TrackPoint" << "Evdev Wheel Emulation" << "1";
        wheelButtonArgs<< "set-prop" << "TPPS/2 IBM TrackPoint" << "Evdev Wheel Emulation Button" << "2";
        wheelAxisArgs<< "set-prop" << "TPPS/2 IBM TrackPoint" << "Evdev Wheel Emulation Axes" << "6 7 4 5";

        if (runMode < 2) {
            trackpointEvdevConf.open(QIODevice::WriteOnly | QIODevice::Truncate);
            QTextStream stream(&trackpointEvdevConf);
            stream << "Section \"InputClass\"\n";
            stream << "\tIdentifier\t\"trackpoint catchall\"\n";
            stream << "\tMatchIsPointer\t\"true\"\n";
            stream << "\tMatchProduct\t\"TrackPoint|DualPoint Stick\"\n";
            stream << "\tMatchDevicePath\t\"/dev/input/event*\"\n";
            stream << "\tOption\t\"Emulate3Buttons\"\t\"true\"\n";
            stream << "\tOption\t\"EmulateWheel\"\t\"1\"\n";
            stream << "\tOption\t\"EmulateWheelButton\"\t\"2\"\n";
            stream << "\tOption\t\"XAxisMapping\"\t\"6 7\"\n";
            stream << "\tOption\t\"YAxisMapping\"\t\"4 5\"\n";
            stream << "EndSection";
            trackpointEvdevConf.close();
        }
    }
    else if (!ui->check_Scrolling->isChecked()) {
        qDebug() << "check_Scroling: Un-checked";
        wheelEmuArgs<< "set-prop" << "TPPS/2 IBM TrackPoint" << "Evdev Wheel Emulation" << "0";
        wheelButtonArgs<< "set-prop" << "TPPS/2 IBM TrackPoint" << "Evdev Wheel Emulation Button" << "3";
        wheelAxisArgs<< "set-prop" << "TPPS/2 IBM TrackPoint" << "Evdev Wheel Emulation Axes" << "0 0 4 5";

        if (runMode < 2) {
            trackpointEvdevConf.remove();
        }
    }

    runXinput.start(xinputCommand, wheelEmuArgs);
    runXinput.waitForFinished(); // sets current thread to sleep and waits for process to end
    runXinput.start(xinputCommand, wheelButtonArgs);
    runXinput.waitForFinished(); // sets current thread to sleep and waits for process to end
    runXinput.start(xinputCommand, wheelAxisArgs);
    runXinput.waitForFinished(); // sets current thread to sleep and waits for process to end
}
